### DISCUSSION TITLE
circuit::Experssion 可以写得很复杂吗？

### DISCUSSION THREAD

**Discussion by voidcenter** [z2o-k7e/halo2-step-by-step#11] [Q&A]
在 is_Zero 这个例子里用到了[Expression](https://github.com/icemelon/halo2-examples/blob/master/src/is_zero.rs#L10)。在表达门约束的时候[用Expression做了运算](https://github.com/icemelon/halo2-examples/blob/master/src/is_zero.rs#L45-L46)。按我理解这是证明路径的一部分，所以这个Experssion应该是不能太复杂的，否则无法被Plonk的最基本的约束格式支持？
`qL * a + qR * b + qO * c + qM  * a * c + qC = 0`

按我理解Halo 2的约束最终都要表达成上面这个样子，那Halo 2内部是有一样自动把 Expression 转换成上面这个格式的机制吗？那需不需要更多的 column 才行？

Expression的定义在[这里](https://github.com/zcash/halo2/blob/main/halo2_proofs/src/plonk/circuit.rs#L489)，注释里写了这个表达式是低阶的，但是定义是递归的，不知道怎么阻止高阶。

我试着把 is_Zero 里面的表达式改成了这样，运行 `MockProver::run` 也没有报错  （是因为 MockProver 只运行，不证明吗？）
```
            is_zero_expr = Expression::Constant(F::ONE) - value.clone() * value_inv 
                + value.clone() * value.clone() * value.clone() * value.clone() * value.clone();

```

**Comment by voidcenter** [comment]
Chatgpt 的回答

<img width="725" alt="Screenshot 2023-10-17 at 2 12 58 PM" src="https://github.com/zkp-co-learning/halo2-step-by-step/assets/33541165/e3dfa4f7-091f-4776-911b-7b0ffd6ba778">

**Reply by voidcenter** [reply]
但 Chatgpt 还是没有回答 halo 2 会不会自己增加 column 来翻译表达式

**Comment by Demian101** [comment]
MockProver 会帮你检查约束的正确性，一般约束正确就意味着可以正常 generate proof.

----

https://github.com/zcash/halo2/blob/main/halo2_proofs/src/plonk/circuit.rs#L488-L509

```rust
/// Low-degree expression representing an identity that must hold over the committed columns.
#[derive(Clone)]
pub enum Expression<F> {
    /// This is a constant polynomial
    Constant(F),
    /// This is a virtual selector
    Selector(Selector),
    /// This is a fixed column queried at a certain relative location
    Fixed(FixedQuery),
    /// This is an advice (witness) column queried at a certain relative location
    Advice(AdviceQuery),
    /// This is an instance (external) column queried at a certain relative location
    Instance(InstanceQuery),
    /// This is a negated polynomial
    Negated(Box<Expression<F>>),
    /// This is the sum of two polynomials
    Sum(Box<Expression<F>>, Box<Expression<F>>),
    /// This is the product of two polynomials
    Product(Box<Expression<F>>, Box<Expression<F>>),
    /// This is a scaled polynomial
    Scaled(Box<Expression<F>>, F),
}
```

你观察的非常正确，因为 `expression` 的定义是递归的，类似 Rust 中的链表，是一个自指的结构，这意味着你可以无限去构建高阶的乘法。但是整个表达式的约束必须小于整个电路所能表达的阶数

----

custom gate(Ultra PLONK 引入) 给了 PLONK 这种高阶的表现力，用来处理复杂地逻辑，这也是比 R1CS、Circom 强大的地方

关于 “需不需要更多的 column 才行” 可能需要观察下 custom gate 在构建时，如何在电路中布局 region

**Reply by mdj33** [reply]
我也有同样的问题，现在知道custom gate不宜阶数过大，现在一般最大是4阶，如果需要更复杂的表达，需要拆成多个express，但是从数学原理上还是比较模糊， 比如类似plonk里面的 qM*a*b, 这里插值成多项式时候，就会有3n次的多项式，由于超出了FFT支持的n次，需要把3n次的多项式拆成3个n次的来算，如果custom gate到4n，5n，就要拆成4个n次，5个n次多项式来算，这样会增加FFT的计算时间，但是个人理解的比较模糊，没有真正从数学上理解清楚，开会时候可以请教下老师

**Reply by voidcenter** [reply]
ok, 我再理解理解，感觉还不是特别懂

如果我要写一个复杂的计算，我该用电路写出来呢，还是用expression直接写出来呢？
